### DISCUSSION TITLE
Measuring Latency by leveraging (hacking) the header

### DISCUSSION THREAD

**Discussion by forsythac** [jacktrip/jacktrip#654] [Ideas]
We have an idea to measure round trip latency, but we need to jam an additional ID into the header somehow...

Since timestamp is being sent by client, server could respond with the same timestamp in order to compare the two and get round trip latency.

Problem is that we need to be able to match the returned timestamps/packets to the ones that were sent, we're thinking that the server could return `SeqNumber` in the header somehow to facilitate the comparison.

But there's no space in the header- @dyfer wants to abuse the sample rate (or other field that's not checked after startup). Server could send back the exact packet but with Sample rate = ~~-1~~ 0 (edit) or some such thing to indicate that it's a bogus audio packet that should only be used for latency measurements. Man, that's hacky but would probably work 😂 

@cchafe please save us from ourselves. @ntonnaett just said "it's the cleanest solution I can think of". I think we might do it

**Comment by cchafe** [comment]
Sorry I missed the meeting -- just got back to internet for a couple
days while roaming the NW forests.

Another way to go would be the trick that's used for the exit packet.
Here's an "echo request" version of that with it's own sequence:
    int seq = 0; // increment me each time
    int controlPacketSize = 62; // where exit packet size = 63
    std::cout << "sending echo request packet " << seq << std::endl;
    QByteArray echoBuf;
    echoBuf.resize(controlPacketSize);
    echoBuf.fill(seq,controlPacketSize);
    writeDatagram(echoBuf, serverHostAddress, mPeerUdpPort);
Just needs an "echo reply" and timing measurement to make it a (quiet)
audio stream "ping."

Chris

**Reply by ntonnaett** [reply]
The good thing about the approach we discussed in the meeting is that it doesn't introduce any overhead and it really uses the audio packets for measurements.
